clock.go

```diff
@@ -102,9 +103,41 @@ func (clock *Clock) Advance(d time.Duration) {
 	clock.mu.Lock()
 	defer clock.mu.Unlock()
 	clock.now = clock.now.Add(d)
+	if len(clock.waiting) == 0 {
+		logger.Debugf("advancing a clock that has nothing waiting")
+	}
 	clock.triggerAll()
 }
 
+// WaitAdvance functions the same as Advance, but only if there is n timers in
+// clock.waiting. This came about while fixing lp:1607044 intermittent
+// failures.  It turns out that testing.Clock.Advance might advance the time
+// and trigger notifications before triggers are set. So we wait a limited time
+// 'w' for 'n' timers to show up in clock.waiting, and if they do we advance
+// 'd'.
+func (clock *Clock) WaitAdvance(d, w time.Duration, n int) error {
+	if w == 0 {
+		w = time.Second
+	}
+	pause := w / 10
+	for i := 0; i < 10; i++ {
+		if clock.hasNWaiters(n) {
+			clock.Advance(d)
+			return nil
+		}
+		time.Sleep(pause)
+	}
+	return errors.Errorf("no timers added after waiting %s", w.String())
```

Comment:
This needs to be updated for the `n` waiters change. It's probably worth saying how many waiters there were. Of course, then you'll get the irritating situation where one shows up between the last check and the report: "waiting for 10 timers, only 10 seen".